Add auxvar and quadrature options to ContactAction

TheBEllis opened the pull request:
Added two more input parameters to the ContactAction
- quadrature, which allows the user to change the quadrature order used with Mortar formulations
- add_aux_var, which allows the user to prevent the generation of `contact_pressure`, `penetration`, `nodal_area` and  `mortar_tangent` aux variables when using mortar

<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->
- Being able to change the quadrature order from the ContactAction is a nice quality of life improvement.
- For large mortar problems, being able to turn off the auxiliary variables required for penalty and kinematic formulations can save a substantial amount of memory. 

## Design
<!--A concise description (design) of the enhancement.-->
- I have wrapped the generation of the aux variables in an `if` statement that will only be skipped if the formulation is set to `mortar` and add_aux_vars is set to `false`. 
- By default add_aux_vars is set to true, and if the user tries to use it with a formulation that isn't `mortar` they will receive an error.
- The quadrature parameter is passed on to the Mortar constraints that can consume it.
- If the user tries to set quadrature with a formulation that isn't mortar they will receive an error. 

## Impact
<!--Will the enhancement change existing APIs or add something new?-->
- Could help enable larger mortar formulation solves through reduced memory footprint.

Edit: 
Closes #30208.

milljm commented:
You'll need to adjust the base branch you are attempting to merge into (change to `next` instead of `devel`)

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request

milljm commented:
Looks like [Precheck](https://civet.inl.gov/job/2783064/) failed, for the following:

```pre
##########################################################################
INFO: Your patch contains no trailing whitespace.
INFO: Your patch contains no proprietary or classified keywords.
INFO: Your patch contains no files without newlines before EOF.
INFO: Your patch contains no bad executable files.
INFO: Your patch contains no banned functions.
INFO: Your patch contains no banned keywords.
INFO: Style check disabled
INFO: Your patch contains no unicode characters.
INFO: Your patch contains no old style C++ include guards.
INFO: Your patch contains windows line endings.
##########################################################################


##########################################################################
ERROR: Your patch does not contain a valid ticket reference! (i.e. #1234)
Merge branch 'ContactActionQuadrature' of https://github.com/TheBEllis/moose into test
Added add_aux_vars to initialiser list
Add auxvar and quadrature options to ContactAction

ERROR: MOOSE prefers two spaces instead of tabs. The following files contain tab characters:
	modules/contact/src/actions/ContactAction.C
##########################################################################
```

Seems you'll need to create an Issue (like a feature request), and then amend your commit with that reference ticket.

TheBEllis commented:
I've created and linked a relevant issue now, hopefully the checks will run fine!

milljm commented:
You just need to include that in your commit message :) Civet is a real stickler about that:

```bash
git commit --ammend

"""
This adds the following features, etc etc

Closes #30208
"""
```

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2783557/), step Clang format on dddc2f8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30198/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30198/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 48c61aaa5b25379d768e22dede9ee8d18641fdd2`

TheBEllis commented:
> You just need to include that in your commit message :) Civet is a real stickler about that:
I re-read your first comment and realised I had been foolish:)

milljm commented:
With Precheck passing, I'll hand this off to one of our developers for approval to run on our clusters.

Cheers!

GiudGiud commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -383,6 +400,14 @@ ContactAction::ContactAction(const InputParameters & params)
+      paramError("add_aux_vars",
+                 "The 'add_aux_vars' option can only be used with the 'mortar' formulation. "
+                 "nodal_area, penetration and contact_pressure variables are not optional with "
+                 "kinematic and penalty formulations.x");
```
```suggestion
                 "kinematic and penalty formulations.");
```

GiudGiud commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -273,6 +273,22 @@ ContactAction::validParams()
+  params.addParam<MooseEnum>(
+      "quadrature",
+      MooseEnum("DEFAULT FIRST SECOND THIRD FOURTH", "DEFAULT"),
+      "Quadrature rule to use on mortar segments. For 2D mortar DEFAULT is recommended."
```
this is the order rather than the rule

let's rename to quadrature_order

GiudGiud commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -273,6 +273,22 @@ ContactAction::validParams()
+      "exact integration of QUAD mortar faces requires SECOND order quadrature for FIRST variables "
+      "and FOURTH order quadrature for SECOND order variables.");
+
+  params.addParam<bool>("add_aux_vars",
```
I would rename to generate_output and let the developers handle the minimum / maximum levels of outputs each formulation can handle

GiudGiud replied:
generate_output is also the name used in the main SM action

GiudGiud commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -273,6 +273,22 @@ ContactAction::validParams()
+  params.addParam<bool>("add_aux_vars",
+                        true,
+                        "For Mortar contact formulation problems, does the user still wish to "
+                        "generate aux-variables necessary for penalty and kinematic formulations."
```
```suggestion
                        "generate auxiliary variables necessary for penalty and kinematic formulations."
```

GiudGiud commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -273,6 +273,22 @@ ContactAction::validParams()
+                        true,
+                        "For Mortar contact formulation problems, does the user still wish to "
+                        "generate aux-variables necessary for penalty and kinematic formulations."
+                        "Turning this to false can be useful if the user is running out of memory "
+                        "on mortar contact solves using a large mesh");
```
```suggestion
);
```
could be other reasons and someone running out of memory will not find this advice anyway

GiudGiud commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -469,165 +494,172 @@ ContactAction::act()
-    // Create auxiliary kernels for each contact pairs
-    for (const auto & contact_pair : _boundary_pairs)
-    {
+  if (!(_formulation == ContactFormulation::MORTAR && !_add_aux_vars))
```
```suggestion
  if (_add_aux_vars || _formulation != ContactFormulation::MORTAR)
```
easier to parse

GiudGiud reviewed (changes requested):
@maxnezdyur would you want to review?

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/2783654/) on 4d81a5a : invalidated by @GiudGiud

unrelated failure

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2783639/), step Docs: sync website on 4d81a5a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30198/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2783655/), step Generate coverage on 4d81a5a wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>48c61a</th>
      <th colspan=4>#30198 4d81a5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.31%</td>
      <td>90.30%</td>
      <td>-0.01%</td>
      <td>97.17%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4874</td>
      <td>4885</td>
      <td>+11</td>
      <td>103</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>523</td>
      <td>525</td>
      <td>+2</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30198/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30198/coverage/contact)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30198/coverage/xfem)

</details>



This comment will be updated on new commits.

maxnezdyur commented:
@TheBEllis 
Can you add a few more options for mortar contact? The action can't set some options that can be useful.

1. `minimum_projection_angle`  
2. `debug_mesh`

GiudGiud commented:
Just a note that adding new options is not mandatory.

But now that the additional need has been expressed, let's at least make sure the PR does not preclude meeting it

maxnezdyur commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -383,6 +400,14 @@ ContactAction::ContactAction(const InputParameters & params)
                  "The 'mortar_dynamics' constraint option can only be used with the 'mortar' "
                  "formulation and in dynamic simulations using Newmark-beta");
+    else if (params.isParamSetByUser("quadrature"))
+      paramError("quadrature",
+                 "The 'quadrature' option can only be used with the 'mortar' formulation.");
```
To prevent confusion, the quadrature option can also be used with `mortar_penalty`. I would call that out.

maxnezdyur reviewed (changes requested):
This adds a QoL improvement that I will be using frequently! Just had a single minor comment.

jbadger95 commented:
Hi Alex, triangle quadratures exactly integrate polynomials with total
order; e.g. x^2 + xy + y^2 + LOT. Quad quadrature rules exactly integrate
polynomials with tenor order; e.g. x^2y^2 + LOT. The quadrature is
evaluating inner products so the 2n+1 is essentially split between the two
functions being integrated. Maybe it’s possible to devise nonstandard
quadrature schemes for triangles and in any case you can often get away
with less since the mortar segment mesh is finer but I think the comment is
correct as written; exact integration with standard quadratures do require
that. Anyway, that’s the motivation on the comment, you probably have more
insight into practice, etc.

On Mon, Mar 31, 2025 at 3:16 PM Alex Lindsay ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In modules/contact/src/actions/ContactAction.C
> <https://github.com/idaholab/moose/pull/30198#discussion_r2021684773>:
>
> > +      "While DEFAULT quadrature order is typically sufficiently accurate, "
> +      "exact integration of QUAD mortar faces requires SECOND order quadrature for FIRST variables "
> +      "and FOURTH order quadrature for SECOND order variables.");
>
> What quadrature order are you seeing? The default quadrature order should
> be 2*n + 1, where n is the order of the FE basis
> ------------------------------
>
> In modules/contact/src/actions/ContactAction.C
> <https://github.com/idaholab/moose/pull/30198#discussion_r2021685444>:
>
> > @@ -273,6 +273,22 @@ ContactAction::validParams()
>        "Whether to use the Petrov-Galerkin approach for the mortar-based constraints. If set to "
>        "true, we use the standard basis as the test function and dual basis as "
>        "the shape function for the interpolation of the Lagrange multiplier variable.");
> +
> +  params.addParam<MooseEnum>(
> +      "quadrature",
> +      MooseEnum("DEFAULT FIRST SECOND THIRD FOURTH", "DEFAULT"),
> +      "Quadrature rule to use on mortar segments. For 2D mortar DEFAULT is recommended."
> +      "For 3D mortar, QUAD meshes are integrated using triangle mortar segments. "
> +      "While DEFAULT quadrature order is typically sufficiently accurate, "
> +      "exact integration of QUAD mortar faces requires SECOND order quadrature for FIRST variables "
> +      "and FOURTH order quadrature for SECOND order variables.");
> +
> +  params.addParam<bool>("add_aux_vars",
> +                        true,
> +                        "For Mortar contact formulation problems, does the user still wish to "
>
> ⬇️ Suggested change
>
> -                        "For Mortar contact formulation problems, does the user still wish to "
> +                        "For mortar contact formulation problems, does the user still wish to "
>
> ------------------------------
>
> In modules/contact/src/actions/ContactAction.C
> <https://github.com/idaholab/moose/pull/30198#discussion_r2021707427>:
>
> > +      "While DEFAULT quadrature order is typically sufficiently accurate, "
> +      "exact integration of QUAD mortar faces requires SECOND order quadrature for FIRST variables "
> +      "and FOURTH order quadrature for SECOND order variables.");
>
> Ah I see this comment is a copy and paste from the doc string in
> MortarConstraintBase. I don't like the comment as currently written.
> Because the quadrature order we end up applying is actually 2*n + 1 where n
> is the order supplied to this parameter. So to describe what is currently
> happening in MOOSE, if someone supplies SECOND to this parameter, you
> will actually get FIFTH for the order value supplied to the quadrature
> constructor. And if someone supplies FOURTH, then the quadrature order
> will end up being NINTH.
>
> I think that instead of duplicating parameters and doc strings which we
> directly forward from the action to the constraint, we should create some
> common params static function that returns these duplicates. Or I think
> maybe @GiudGiud <https://github.com/GiudGiud> created APIs for doing
> these parameter transfers for physics?
>
> Finally, @jbadger95 <https://github.com/jbadger95> or @roystgnr
> <https://github.com/roystgnr> do we really need NINTH quadrature for
> exactly integrating second order polynomial bases on a quad face that has
> been decomposed into two tris?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/30198#pullrequestreview-2730372749>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJACZHVVKC4PUDDYABRIJZL2XGPBXAVCNFSM6AAAAABZ4Z2PACVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDOMZQGM3TENZUHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

lindsayad commented:
That's clear. Thanks @jbadger95 for your quick reply! Hope you're doing well 😄

roystgnr commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -273,6 +273,22 @@ ContactAction::validParams()
+      MooseEnum("DEFAULT FIRST SECOND THIRD FOURTH", "DEFAULT"),
+      "Quadrature rule to use on mortar segments. For 2D mortar DEFAULT is recommended."
+      "For 3D mortar, QUAD meshes are integrated using triangle mortar segments. "
+      "While DEFAULT quadrature order is typically sufficiently accurate, "
+      "exact integration of QUAD mortar faces requires SECOND order quadrature for FIRST variables "
+      "and FOURTH order quadrature for SECOND order variables.");
```
Define "decomposed into two tris"?  We're using a composition of two quadrature rules on triangles to integrate biquadratic functions?  The function space is no biquadratic, we've split the element and now it's piecewise quadratic?

Anyway: you can't *exactly* integrate biquadratic polynomial bases on a general quad face with *any* quadrature order: you're wanting to integrate φᵢφⱼ, which is biquartic, so you'd think `FOURTH` would do it, but when you pull it back to the master element where your quadrature rule was defined you tack on a |∂x/∂ξ| term that isn't polynomial at all.  So we content ourselves with integrating, not exactly, but closely enough that the error term is swamped by our numerical approximation error.  For general nonlinear PDEs 2p+1 is usually safe (though I don't recall the conditions guaranteeing that), and with no nonlinearity in the physical-space integral (and an invertible x(ξ); we're hoping to deal with distorted mappings, not singular ones) 2p should be fine.

lindsayad replied:
Decomposed into two tris was very inaccurate. We decompose a given face element into `n_sub_elem()`, then we linearize it. We do this for both secondary and primary faces, and then we perform projections with these linearized sub-elements and then triangulate the result. @jbadger95 wrote up some nice documentation (with pictures!) [here](https://mooseframework.inl.gov/source/constraints/AutomaticMortarGeneration.html).

So we apply a quadrature rule on the TRI3 elements which is of order 2*n + 1, where n is the order that the user passes to this parameter, e.g. the user parameter n corresponds to the order of the FE basis, *not* the quadrature.

We take the quadrature points on these TRI3 elements, map them back onto the possibly non-planar sub-elements, and then finally back onto the original possibly quadratic face elements. We then evaluate our finite element bases and variables with the mapped face element quadrature. Is that clear?

So this is why @jbadger95 is saying that you need to square the FE basis order ($n \mapsto n^2$) when your face elements are quads because you will ultimately be mapping a quadrature on TRI3 (linear) back to these quads (bilinear, biquadratic, etc.)

lindsayad replied:
Once we're all on the same page, I'm probably going to add some more documentation to the repo capturing this conversation

lindsayad commented:
we could probably finish this if @TheBEllis would prefer us to

GiudGiud commented:
Yeah we should. @lindsayad want to take it on and I review or the opposite

lindsayad committed:
Add quadrature option to ContactAction

Added one more input parameter to the ContactAction
- quadrature, which allows the user to change the quadrature order used with Mortar formulations

Closes #32095

lindsayad committed:
Ignore Nvidia profiler output

lindsayad committed:
Add more doc about mortar quadrature parameter

Also transfer param to ContactAction instead of copying so that
we don't have to do things like update doc strings in two places

lindsayad committed:
Add second order quadrature test
